solidity/contracts/hooks/CctpHook.sol

```diff
@@ -0,0 +1,130 @@
+        tokenMessenger = _tokenMessenger;
+        childHook = _childHook;
+    }
+
+    /**
+     * @notice Emitted when the Hyperlane domain to Circle domain mapping is updated.
+     * @param hyperlaneDomain The Hyperlane domain.
+     * @param circleDomain The Circle domain.
+     */
+    event DomainAdded(uint32 indexed hyperlaneDomain, uint32 circleDomain);
+
+    /// @notice Hyperlane domain => Circle domain.
+    /// ATM, known Circle domains are Ethereum = 0, Avalanche = 1, Optimism = 2, Arbitrum = 3.
+    /// Note this could result in ambiguity between the Circle domain being
+    /// Ethereum or unknown.
+    mapping(uint32 => uint32) public hyperlaneDomainToCircleDomain;
+
+    /**
+     * @notice Adds a new mapping between a Hyperlane domain and a Circle domain.
+     * @param _hyperlaneDomain The Hyperlane domain.
+     * @param _circleDomain The Circle domain.
+     */
+    function addDomain(
+        uint32 _hyperlaneDomain,
+        uint32 _circleDomain
+    ) external onlyOwner {
+        hyperlaneDomainToCircleDomain[_hyperlaneDomain] = _circleDomain;
+
+        emit DomainAdded(_hyperlaneDomain, _circleDomain);
+    }
+
+    function hookType() external pure override returns (uint8) {
+        return uint8(IPostDispatchHook.Types.CCTP);
+    }
+
+    /// @inheritdoc AbstractPostDispatchHook
+    function _quoteDispatch(
+        bytes calldata metadata,
+        bytes calldata message
+    ) internal view override returns (uint256 fee) {
+        return igp.quoteDispatch(metadata, message);
+    }
+
+    /// @inheritdoc AbstractPostDispatchHook
+    function _postDispatch(
+        bytes calldata metadata,
+        bytes calldata message
+    ) internal override {
+        // Wrapped token is encoded into the metadata, see TokenBridgeCctp
+        address token = address(
+            bytes20(TokenMessage.metadata(message.body())[0:20])
+        );
+        uint256 amount = TokenMessage.amount(message.body());
+
+        IERC20(token).approve(address(tokenMessenger), amount);
+
+        uint32 circleDomain = hyperlaneDomainToCircleDomain[
+            message.destination()
+        ];
+
+        bytes32 recipient = message.recipient();
+
+        if (tokenMessenger.messageBodyVersion() == 0) {
+            // CCTP v1
+            tokenMessenger.depositForBurn(
+                amount,
+                circleDomain,
+                recipient,
+                token
+            );
+        } else if (tokenMessenger.messageBodyVersion() == 1) {
+            // CCTP v2
+            ITokenMessengerV2(address(tokenMessenger)).depositForBurn(
+                amount,
+                circleDomain,
+                recipient,
+                token,
+                bytes32(0),
+                CCTP_V2_DEFAULT_MAX_FEE,
+                CCTP_V2_DEFAULT_MIN_FINALITY_THRESHOLD
+            );
+        }
+
+        uint256 fees = igp.quoteDispatch(metadata, message);
+        igp.postDispatch{value: fees}(metadata, message);
```

Comment:
this call doesnt really make sense to me

Reply:
the `message` used here will be the warp route message from the `TokenBridgeCCTP` contract

Reply:
as written this hook cannot be used with any other message sender

Reply:
an alternative is to not express this logic as a hook at all and lift this up into the `TokenBridgeCCTP.transferRemote` logic